### PR TITLE
feat(platform-browser): Expose EventManagerPlugin in the public api

### PR DESCRIPTION
**Pull request by JeanMeche**
This commit exposes the `EventManagerPlugin` abstract class to the public API.
It provides a documentation with a usecase and a basic implementation describing the possibilities

Rationale for opening to the public API :

1. The `EVENT_MANAGER_PLUGINS` token is already public
2. The documentation provides a usefull usecase
3. Plugins can already be implemented by following the same API
4. This does not increase the API surface because of 3.
5. This topic gained some visibility in the last few days in the twitter community and seems to be not so widely known, so this helps also for the visibility ! (Also wanted to share the usescases implemented by [@tinkoff/ng-event-plugins](https://github.com/Tinkoff/ng-event-plugins))

## PR Type
What kind of change does this PR introduce?

- [x] Feature

## Does this PR introduce a breaking change?

- [x] No

### CODE REVIEW

File: packages/platform-browser/src/dom/events/event_manager.ts [hunk 1]
```diff
@@ -9,7 +9,7 @@
 import {Inject, Injectable, InjectionToken, NgZone} from '@angular/core';
 
 /**
- * The injection token for the event-manager plug-in service.
+ * The injection token for the for plugins of the `EventManager` service.
```
Comment (pweyrich) [inline packages/platform-browser/src/dom/events/event_manager.ts:12]:
```suggestion
 * The injection token for plugins of the `EventManager` service.
```

**Comment by JeanMeche** [comment]
Can I get the [aio: preview](https://github.com/angular/angular/pulls?q=is%3Aopen+is%3Apr+label%3A%22aio%3A+preview%22) tag ?

**Comment by waterplea** [comment]
Super glad to see movement in this direction 👍

Just wanted to mention these two issues slightly impeding wide adoption of custom plugins:

https://github.com/angular/angular/issues/37850
https://github.com/angular/angular/issues/40553

File: goldens/public-api/platform-browser/index.md [hunk 2]
```diff
@@ -92,6 +92,18 @@ export class EventManager {
     static ɵprov: i0.ɵɵInjectableDeclaration<EventManager>;
 }
 
+// @public
+export abstract class EventManagerPlugin {
+    abstract addEventListener(element: HTMLElement, eventName: string, handler: (event: Event) => void): Function;
+    // (undocumented)
+    manager: EventManager;
```
[the remark this answers is not in the document]
Comment (JeanMeche) [reply goldens/public-api/platform-browser/index.md:99]: Great idea, that property bothered me also !

[the remark this answers is not in the document]
Comment (waterplea) [reply goldens/public-api/platform-browser/index.md]: Without that property, the example doesn't work, as well as the broader concept of unintrusive plugins that do their part and delegate further handling back to the manager which is heavily used by [@tinkoff/ng-event-plugins](https://github.com/Tinkoff/ng-event-plugins).

[the remark this answers is not in the document]
Comment (JeanMeche) [reply goldens/public-api/platform-browser/index.md]: The problem of accessing the manager can be solved by injecting it ! 

```typescript
export class StopEventPlugin {
  constructor(private injector: Injector) {}

  supports(event: string): boolean {
    return event.endsWith('.stop');
  }

  addEventListener(element: HTMLElement, eventName: string, handler: (event: Event) => void): Function {
    const eventWrapper = (event: Event) => {
      event.stopPropagation();
      handler(event);
    };

    // striping the stop modifier from the event name;
    const nonModifiedEventName = eventName.replace(/\.stop$/, '');

    const manager = this.injector.get(EventManager);
    return manager.addEventListener(element, nonModifiedEventName, eventWrapper);
  }
}
```

What do you think ?  Should we revert the breaking change or suggest to inject the manager if it's needed ?

[the remark this answers is not in the document]
Comment (waterplea) [reply goldens/public-api/platform-browser/index.md]: Personally, I don't think switching from property assignment by `EventManager` to the entire `Injector` to avoid cyclic dependency is much better. Sounds like for plugins to have direct access to the manager is reasonable.

[the remark this answers is not in the document]
Comment (atscott) [reply goldens/public-api/platform-browser/index.md:99]: > Sounds like for plugins to have direct access to the manager is reasonable.

Having access to the manager isn't really the issue, but rather the awkwardness of having a public API where the some other service assigns a value to a user's publicly writable property after the service is created.

**Comment by atscott** [comment]
@JeanMeche Can you separate this in to two PRs? One that simply adds the `@publicApi` annotation to `EventManagerPlugin` and another that adds the documentation? For the public API change, you can simply cite the existing token: https://angular.io/api/platform-browser/EVENT_MANAGER_PLUGINS. The exposed type of the token should definitely be in the public API.

File: goldens/public-api/platform-browser/index.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (waterplea) [reply goldens/public-api/platform-browser/index.md:99]: Maybe use the same approach control value accessors have and add `registerManager(manager: EventManager): void`?

**Comment by JeanMeche** [comment]
@atscott I've reduced the scope of the PR to adding `EventManagerPlugin` to the public API.

Do you have any clue why `getFactoryOf` has been added to the goldens ?

**Comment by atscott** [comment]
> @atscott I've reduced the scope of the PR to adding EventManagerPlugin to the public API.

> Do you have any clue why getFactoryOf has been added to the goldens ?

The PR is still a bit more than adding `EventManagerPlugin` to the public API. There are some other refactors in here that caused the golden files to change and `getFactoryOf` is a result of those refactors. I believe they are there now because you inherit the constructor from the base class rather than having the explicit `super` calls. Could you limit the changes even further?

**Comment by JeanMeche** [comment]
@atscott I wanted to clean a bit the API before making it public, as its always hard to update once it's public.  

* `_doc` in `EventManaerPlugin` is unused so I removed it. 
* I update all `super()` calls as a result. 
* `Function` type for the handler is really vague and I wanted to improve it. 

By restoring the constructor, I got to revert the changes in the golden. 

What do you think of it now ?

**Review by atscott — APPROVED** [review]
> @atscott I wanted to clean up a bit the API before making it public, as its always harder to update once it's public.

Yea, that's a good idea. The changes generally look good to me now.

Note: While abstract, `EventManagerPlugin` is still technically not allowed to be extended by developers due to our "classes are final unless explicitly noted otherwise" policy. You will need to add that developers can extend the class if that's the intent (similar to https://github.com/angular/angular/blob/e40a640dfe54b03bfe917d08098c319b0b200d25/packages/router/src/route_reuse_strategy.ts#L76-L77)

File: packages/platform-browser/src/dom/events/key_events.ts [hunk 3]
```diff
@@ -45,17 +45,12 @@ const MODIFIER_KEY_GETTERS: {[key: string]: (event: KeyboardEvent) => boolean} =
 };
 
 /**
- * @publicApi
  * A browser plug-in that provides support for handling of key events in Angular.
```
Comment (JeanMeche) [inline packages/platform-browser/src/dom/events/key_events.ts:48]: It wasn't public, it's exported as `ɵKeyEventsPlugin`.

File: packages/platform-browser/src/dom/events/dom_events.ts [hunk 4]
```diff
@@ -6,15 +6,14 @@
...
-import {Inject, Injectable} from '@angular/core';
+import {Injectable} from '@angular/core';
 
 import {EventManagerPlugin} from './event_manager';
 
 @Injectable()
 export class DomEventsPlugin extends EventManagerPlugin {
-  constructor(@Inject(DOCUMENT) doc: any) {
-    super(doc);
+  constructor() {
```
Comment (alfaproject) [inline packages/platform-browser/src/dom/events/dom_events.ts:15]: These should be redundant now

File: packages/platform-browser/src/dom/events/key_events.ts [hunk 5]
```diff
@@ -45,17 +45,12 @@ const MODIFIER_KEY_GETTERS: {[key: string]: (event: KeyboardEvent) => boolean} =
...
  */
 @Injectable()
 export class KeyEventsPlugin extends EventManagerPlugin {
-  /**
-   * Initializes an instance of the browser plug-in.
-   * @param doc The document in which key events will be detected.
-   */
-  constructor(@Inject(DOCUMENT) doc: any) {
-    super(doc);
+  constructor() {
```
Comment (alfaproject) [inline packages/platform-browser/src/dom/events/key_events.ts:52]: Redundant now, right?

File: packages/platform-browser/src/dom/events/key_events.ts [hunk 6]
```diff
@@ -188,7 +183,6 @@ export class KeyEventsPlugin extends EventManagerPlugin {
 
   /** @internal */
   static _normalizeKey(keyName: string): string {
-    // TODO: switch to a Map if the mapping grows too much
```
Comment (alfaproject) [inline packages/platform-browser/src/dom/events/key_events.ts:191]: Removing this comment is unrelated to this PR. You probably should revert and actually refactor this to a simple if condition until a switch or map or whatever else is needed because an if condition minifies better

File: packages/platform-browser/src/dom/events/dom_events.ts (hunk 4 above)
Comment (JeanMeche) [reply packages/platform-browser/src/dom/events/dom_events.ts:15]: Actually they aren't redundant. 

Having no constructor pulls `getFactoryOf()` cf my comment above about the goldens & Andrew's answer.

Comment (alfaproject) [reply packages/platform-browser/src/dom/events/dom_events.ts:15]: What kind of sorcery is that? o:

File: packages/platform-browser/src/dom/events/key_events.ts (hunk 6 above)
Comment (JeanMeche) [reply packages/platform-browser/src/dom/events/key_events.ts:191]: Yeah good idea 👍

File: packages/platform-browser/src/dom/events/dom_events.ts (hunk 4 above)
Comment (JeanMeche) [reply packages/platform-browser/src/dom/events/dom_events.ts:15]: I believe its down to the implementation of `ɵɵgetInheritedFactory` : 
https://github.com/angular/angular/blob/8a2739f25034443f3a72d57f983e70c1a9d69722/packages/core/src/render3/di.ts#L722-L751

It calls getFactoryOf() if there is no constructor on the class itself.

**Review by jessicajaniuk — APPROVED** [review]
reviewed-for: public-api

This will definitely need a TGP.

**Review by dylhunn — APPROVED** [review]
reviewed-for: public-api

**Comment by josephperrott** [comment]
@dylhunn replying offline as this is a g3 specific item.

**Comment by dylhunn** [comment]
I will run a TGP on [this CL](https://critique.corp.google.com/cl/530642500) tonight.

**Comment by JeanMeche** [comment]
Any news @dylhunn ?

**Comment by dylhunn** [comment]
TGP is running: https://test.corp.google.com/OCL:560756113:BASE:560894821:1693279799412:6c3fde97

**Comment by dylhunn** [comment]
TGP passed.

**Comment by jessicajaniuk** [comment]
This PR was merged into the repository by commit c5daa6ce776724d44c02cc97f1a349a85cb2a819.
